src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueOfJsonPrimitive.cs

```diff
@@ -0,0 +1,397 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Encodings.Web;
+
+namespace System.Text.Json.Nodes
+{
+    internal static class JsonValueOfJsonPrimitive
+    {
+        internal static JsonValue CreatePrimitiveValue(ref Utf8JsonReader reader, JsonNodeOptions options)
+        {
+            switch (reader.TokenType)
+            {
+                case JsonTokenType.False:
+                case JsonTokenType.True:
+                    return new JsonValueOfJsonBool(reader.GetBoolean(), options);
+                case JsonTokenType.String:
+                    byte[] quotedStringValue;
+
+                    if (reader.ValueIsEscaped)
+                    {
+                        ReadOnlySpan<byte> unescapedValue = reader.GetUnescapedSpan();
+
+                        quotedStringValue = new byte[checked(2 + unescapedValue.Length)];
+                        unescapedValue.CopyTo(quotedStringValue.AsSpan().Slice(1, unescapedValue.Length));
+                    }
+                    else if (reader.HasValueSequence)
+                    {
+                        // Throw if the long to int conversion fails. This can be accommodated in the future if needed by
+                        // adding another derived class that has a backing ReadOnlySequence<byte> instead of a byte[].
+                        int bufferLength = checked((int)(2 + reader.ValueSequence.Length));
+                        quotedStringValue = new byte[bufferLength];
+                        reader.ValueSequence.CopyTo(quotedStringValue.AsSpan().Slice(1, bufferLength - 2));
+                    }
+                    else
+                    {
+                        quotedStringValue = new byte[2 + reader.ValueSpan.Length];
+                        reader.ValueSpan.CopyTo(quotedStringValue.AsSpan(1, reader.ValueSpan.Length));
+                    }
+
+                    quotedStringValue[0] = quotedStringValue[quotedStringValue.Length - 1] = JsonConstants.Quote;
+                    return new JsonValueOfJsonString(quotedStringValue, options);
+                case JsonTokenType.Number:
+                    byte[] numberValue = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan.ToArray();
+                    return new JsonValueOfJsonNumber(new JsonValueOfJsonNumber.JsonNumber(numberValue), options);
+                default:
+                    Debug.Fail("Only primitives allowed.");
+                    ThrowHelper.ThrowJsonException();
+                    return null!; // Unreachable, but required for compilation.
+            }
+        }
+
+        private sealed class JsonValueOfJsonString : JsonValue
+        {
+            private readonly byte[] _value;
+
+            internal JsonValueOfJsonString(byte[] value, JsonNodeOptions? options)
```

Comment:
Nit: use a name suggesting what the field is representing.
```suggestion
            internal JsonValueOfJsonString(byte[] utf8String, JsonNodeOptions? options)
```